src/java/main/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -1054,6 +1054,8 @@ private void sendPing() {
         private boolean saslLoginFailed = false;
 
         private void startConnect() throws IOException {
+            // initializing it for new connection
+            saslLoginFailed = false;
```

Comment:
Only one thread involved in new connection creation. So there will not any chance another thread setting the flag in between.